### PR TITLE
Server: add SO_REUSEPORT socket option if available

### PR DESCRIPTION
**Pull request by kanak**
This allows us to have multiple threads listening on the same port in
Linux. Also add an example demonstrating usage.

### CODE REVIEW

**Comment by kanak** [comment]
Hi @mattgodbolt , please let me know if the changes look good to you and if you'd like further edits.

I need this change to move my application on to seasocks; my websocket commands often result in intensive computation and the easiest way to move things over without rearchitecting everything was to just run N threads per process and let each thread do blocking work if necessary.

**Comment by offa** [comment]
Sorry for the late response. Since `SO_REUSEADDR` is already set, I don't see any drawbacks.

**Review by offa — APPROVED** [review]

**Comment by mattgodbolt** [comment]
Thansk both! Sorry I didn't spot this earlier. I've since fixed up my email filters so I should spot things in future!

File: src/main/c/Server.cpp [hunk 1]
```diff
@@ -175,9 +175,15 @@ bool Server::configureSocket(int fd) const {
     }
     const int yesPlease = 1;
     if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &yesPlease, sizeof(yesPlease)) == -1) {
-        LS_ERROR(_logger, "Unable to set reuse socket option: " << getLastError());
+        LS_ERROR(_logger, "Unable to set reuse address socket option: " << getLastError());
         return false;
     }
+#ifdef SO_REUSEPORT
+    if (setsockopt(fd, SOL_SOCKET, SO_REUSEPORT, &yesPlease, sizeof(yesPlease)) == -1) {
+        LS_ERROR(_logger, "Unable to set resuse port socket option: " << getLastError());
```
Comment (mattgodbolt) [inline src/main/c/Server.cpp:183]: Typo: `resuse` -> `reuse`

**Review by mattgodbolt — COMMENTED** [review]
Other than the typo, this looks good. Thanks for the comments explaining the things one needs to be careful about.

File: src/main/c/Server.cpp (hunk 1 above)
Comment (kanak) [reply src/main/c/Server.cpp]: Fixed.

**Comment by kanak** [comment]
> Sorry for the late response. Since `SO_REUSEADDR` is already set, I don't see any drawbacks.
> 
> @kanak / @mattgodbolt: Does it make sense to set this via option instead of unconditionally?

Hi @offa, 

I think a flag just for `SO_REUSEPORT` might not be a good idea, because `SO_REUSEADDR` by itself on Linux has some annoying limitations ( https://stackoverflow.com/a/14388707) and `SO_REUSEPORT` has some nice performance implications ( https://lwn.net/Articles/542629/ ).

However, a flag to disable both `REUSE` flags altogether might be a good idea? That way you could be sure that there is no other thread/process on the same host that is also listening on the same port.

**Comment by offa** [comment]
I see, so we can add such a flag later (if even necessary).

**Comment by mattgodbolt** [comment]
Agreed we can add flags for either or both later if needed. Thanks!

**Pull request merged**